Add support for Baichuan2

xhaihao opened the pull request:
# What does this PR do?

This PR added the Baichuan2 model (a custom model), including Baichuan2-7B-Chat and Baichuan2-13B-Chat. [1][2]

[1] https://huggingface.co/baichuan-inc/Baichuan2-7B-Chat
[1] https://huggingface.co/baichuan-inc/Baichuan2-13B-Chat

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

xuguangxin commented:
@libinta @sywangyi , pls help review, thanks.

sywangyi commented on optimum/habana/transformers/models/baichuan/quantizer.py:
```diff
@@ -0,0 +1,237 @@
+from bitsandbytes.nn.modules import Params4bit, Int8Params
+import torch
+
+def Params4bitCuda(self, device):
```
cuda? are 4bit and 8bit of baichuan supported, I see you import bitsandbytes. is bitsandbytes support in habana?

xhaihao replied:
Thanks for pointing it out. It doesn't support INT8/4 quantization for baichuan in habana.  I removed the stale code in the new version.

xuguangxin commented:
@libinta @sywangyi @mandy-li, more comment for this?
could you help approve the test workflow on for this commit?

regisss commented:
@xuguangxin Please sync your branch with main and solve the merge conflicts. I'll trigger the CI once this is done :slightly_smiling_face:

xhaihao commented:
@regisss  Done, thx

regisss commented:
@xhaihao Can you add this model in the table of the README and of the docs please?
https://github.com/huggingface/optimum-habana/blob/ba914504bd52b989d05a4e11916aa19b3ce176f4/README.md?plain=1#L192
https://github.com/huggingface/optimum-habana/blob/ba914504bd52b989d05a4e11916aa19b3ce176f4/docs/source/index.mdx?plain=1#L59

There are new merge conflicts to solve also.

xhaihao commented:
> @xhaihao Can you add this model in the table of the README and of the docs please?
> 
> https://github.com/huggingface/optimum-habana/blob/ba914504bd52b989d05a4e11916aa19b3ce176f4/README.md?plain=1#L192
> 
> 
> https://github.com/huggingface/optimum-habana/blob/ba914504bd52b989d05a4e11916aa19b3ce176f4/docs/source/index.mdx?plain=1#L59
> 
> There are new merge conflicts to solve also.

Thanks for the comment, I updated patch and rebased it against the latest main branch.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_1479). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

xhaihao commented on optimum/habana/transformers/models/baichuan/tokenization_baichuan.py:
```diff
@@ -0,0 +1,258 @@
+from shutil import copyfile
+from typing import Any, Dict, List, Optional, Tuple
+
+import sentencepiece as spm
```
@regisss @dsmertin Thanks for catching & fixing the issue. It was my fault because sentencepiece was installed in my environment.
